### DISCUSSION TITLE
How to apply this at scale?

### DISCUSSION THREAD

**Discussion by JP-Dhabolt** [nathro/AutoTransform#75] [Q&A]
First, I wanted to start by thanking you for the work you've put into this project!  I've played around and got this working in a POC for a single repository.  I have questions around how to apply this at scale.  I'll use the following scenario to put context around my questions.

> Github Org `A` has 5 repositories that each have a configuration file that needs to be updated in the same way (simple transformation).

From my understanding of AutoTransform, this would require 5 different schema files (one for each repo, as the repo is part of the schema definition).  Additionally, AutoTransform would need to be part of each repository, as it appears to apply transformations only on the current repo.  The schema repo seems to only come into play when creating a PR, and errors if not referencing the same repo.

Is this understanding correct?

My hope is that something similar to the below diagram would be possible.  There would then be a way to scale out a transformation (defined one time) across N number of repos.
```mermaid
sequenceDiagram
    participant AT as AutoTransformRepo
    participant T as TargetRepo
    AT->>+T: Clones base_branch
    AT->>+AT: Applies Transformation to cloned repo
    AT->>+T: Open PR 

```

If I'm trying to do something that is out of scope for this tool, please let me know.
Thanks!

JP

**Comment by nathro** [comment]
This is definitely not out of the realm of things AutoTransform should be able to solve, it's just not been my main focus. There's a world where you could separate the repo from the schema file and have them passed separately to the run script.

I'd want to think about changes to the structure here that we could use to enable this type of behavior. I suspect it would involve an optional argument to runner scripts that would allow you to override the repo that a schema specifies (at least, that would be the simplest way to implement this). We could also go down the road of fully removing the repo from the schema files and have that supplied separately, maybe via a similar approach to the schema map where we map repo names to objects that you could use.

The biggest question is how do we do this in a way that doesn't make the single repo use case difficult to work with.

I come from a mono repo world and this definitely shows in how I built AutoTransform, heh.

**Comment by nathro** [comment]
As I think about it, what might actually be the cleanest way to do this is via the config. I imagine the changes we would need to make are:
 - Remove repos from schema files
 - Add a repo option to the config
 - Enable the run script to clone the repo
 - Add a way to separate where your files are located so you can have schema files separate from the repos themselves
 
#3 and #4 are where I think we'll run in to the most issues.

**Reply by JP-Dhabolt** [reply]
I like the concept of moving the repo into the config, as that would leave the schema as the recipe for the change, without any direct knowledge of the target of the recipe.

The good news about the 3rd point is that you already have a GH Token that can be used to clone the repo, so I would think it would just be a matter of cloning into a temp directory.  As to the 4th, I think providing an explicit CWD to commands (or as the base for filepaths, etc) should handle most of the heavy lifting.  Of course, the devil is in the details of implementation.

Thanks for confirming that this would fit into the scope of the tool's intent.  I'll see if I can spend some time over the holidays to play around with some modifications.

**Reply by nathro** [reply]
I'm still working on a few things for our deployment at Slack, but I can take a look at this over the holidays. I have some ideas that can get us 90% of the way. I think cloning will be a bit weirder than you're expecting because of the need to support things like ghe, and avoiding have tokens in the URL, because if you have a token in the URL it ends up in the config, which is bad. It's pretty easy to just check it out in your github action so it might just not be worth to do so *shrug*.

On a side note, I'd love to know how you found out about this tool if you don't mind sharing. I haven't been doing a ton of external evangelism so curious how you learned about it.

**Reply by JP-Dhabolt** [reply]
Yeah, I think being able to operate on a different repo is more important than cloning said repo, as that can always be handled externally as you mentioned.
 
I subscribe to the Slack engineering blog RSS feed, and [this popped up about 3 months ago](https://slack.engineering/autotransform-efficient-codebase-modification/).  I thought it was an interesting concept and squirreled it away until I had a use case for it, which happened this week.

**Reply by nathro** [reply]
I think https://github.com/nathro/AutoTransform/pull/78 should solve this problem for you. Now you just need to use the repo_override in the config each of your repos have and use AUTO_TRANSFORM_SCHEMA_MAP_PATH environment variable to specify where the schema map is actually located.

With this you should be able to have your centralized repo that you put any custom components/schemas in, then each repo would just need it's own config. Running github actions off that centralized repo would allow you to checkout your centralized repo as well as a target repo.

**Reply by JP-Dhabolt** [reply]
Awesome, I'll give this a shot today!  Thanks for the quick turnaround on this, I wasn't anticipating anything before the end of the year.

**Reply by nathro** [reply]
It was pretty straightforward so no biggie. It should be up on PyPi as 1.0.8a1 shortly. If you encounter any other issues or have other questions when you try using this let me know :). Excited to see some initial adoption outside Slack.

**Reply by JP-Dhabolt** [reply]
I think there's a few minor issues.  Let me know if you want me to open an issue for these (or submit a PR for them myself).  I'll include them here as well:

* The `Config.merge` does not include the `repo_override`, so it always gets set to `None`.  I modified the installed package locally to include that line and it worked as expected.
* The `EnvironmentConfigFetcher` does not have a section for `AUTO_TRANSFORM_REPO_OVERRIDE`, so this can't be set as an environment variable within a CI pipeline.
* The `AUTO_TRANSFORM_SCHEMA_MAP_PATH` works as expected, but the schema then tries to load based on the current directory (i.e. the repo being modified) as opposed to relative to the `AUTO_TRANSFORM_SCHEMA_MAP_PATH` (i.e. the schema repo).  I am able to set the `target` to an absolute path as a workaround, but this would limit the ability to run locally (for testing) and in a CI runner, as they would very likely have different path prefixes.
* This one is cosmetic vs functional, but in `Config.get_repo_override_from_console`, the `name` passed into the `factory.from_console` is `remote runner` instead of `repo override`, which results in an incorrect prompt.

**Reply by nathro** [reply]
This is what I get for thinking it was trivial, heh. 1, 2, and 4 are pretty straightforward, the 3rd is a bit...stranger. We expect the paths for schemas right now to be relative to where you invoke autotransform from (though you can always use absolute paths). Making those paths relative to where the schema map is would be possible I guess. I'll try something for that.

**Reply by nathro** [reply]
Okay, so

1. https://github.com/nathro/AutoTransform/pull/80
2. https://github.com/nathro/AutoTransform/pull/81
3. Need to think on this for a bit to do it in a non-breaking way
4. https://github.com/nathro/AutoTransform/pull/82

Thanks for catching these for me :)

**Reply by nathro** [reply]
And for 3, I think we'll need to actually do a breaking change here, there's no way around it. https://github.com/nathro/AutoTransform/pull/83 should be a fix for that (while also giving us a nice SchemaMap util class).

**Reply by JP-Dhabolt** [reply]
Thanks again for the fast turnaround on this.  I'll kick the tires for this on Monday.  Have a good Thanksgiving!

**Reply by nathro** [reply]
I released 1.1.0a1 for you to kick the tires on :)

**Reply by JP-Dhabolt** [reply]
Thanks!  This change works as expected for my use case, using the `AUTO_TRANSFORM_REPO_OVERRIDE` environment variable.

**Reply by nathro** [reply]
Glad to hear it, if you encounter any other issues or need some help with something let me know.
